### PR TITLE
[WIP][Community Pipeline]  InstaFlow! One-Step Stable Diffusion with Rectified Flow 

### PR DESCRIPTION
**Pull request by ayushtues**
Following https://github.com/huggingface/diffusers/issues/6037

### Model/Pipeline/Scheduler description

InstaFlow is an ultra-fast, one-step image generator that achieves image quality close to Stable Diffusion, significantly reducing the demand of computational resources. This efficiency is made possible through a recent [Rectified Flow](https://github.com/gnobitab/RectifiedFlow) technique, which trains probability flows with straight trajectories, hence inherently requiring only a single step for fast inference.

InstaFlow has several advantages:

Ultra-Fast Inference: InstaFlow models are one-step generators, which directly map noises to images and avoid multi-step sampling of diffusion models. On our machine with A100 GPU, the inference time is around 0.1 second, saving ~90% of the inference time compared to the original Stable Diffusion.
High-Quality: InstaFlow generates images with intricate details like Stable Diffusion, and have similar FID on MS COCO 2014 as state-of-the-art text-to-image GANs, like [StyleGAN-T](https://github.com/autonomousvision/stylegan-t).
Simple and Efficient Training: The training process of InstaFlow merely involves supervised training. Leveraging pre-trained Stable Diffusion, it only takes 199 A100 GPU days to get InstaFlow-0.9B.

image

### Open source status

- [X] The model implementation is available.
- [X] The model weights are available (Only relevant if addition is not a scheduler).

### Provide useful links for the implementation

Code & Weights - https://github.com/gnobitab/InstaFlow
HF Space - https://huggingface.co/spaces/XCLiu/InstaFlow

### TO-DO

- [ ] Pipeline code
- [ ] Documentation
- [ ] Tests

### CC
@sayakpaul @gnobitab

### CODE REVIEW

**Comment by sayakpaul** [comment]
Let's fix the styling problems.

**Comment by sayakpaul** [comment]
Also, have you tried the pipeline yourself? Have you experienced any significant speedup? Could you maybe report some results (possibly with the generated images)?

**Comment by ayushtues** [comment]
Will try adding the LoRa results they show, since they have a better quality

**Comment by sayakpaul** [comment]
Gentle ping @ayushtues.

**Comment by ayushtues** [comment]
Will share an update soon, was away on a trip for a while

**Comment by ayushtues** [comment]
While I am not very familiar with benchmarking timings, at 32 bit precision, on a V100, I was able to generate images at around 0.03 secs

```python
from examples.community.instaflow_one_step import InstaFlowPipeline
import torch


pipe = InstaFlowPipeline.from_pretrained("XCLIU/instaflow_0_9B_from_sd_1_5", torch_dtype=torch.float32, cache_dir="./cache", requires_safety_checker=False)
pipe.do_lora()
pipe.to("cuda")  ### if GPU is not available, comment this line
prompt = "A hyper-realistic photo of a cute cat."

image_list = []
for i in range(10):
    images = pipe(prompt=prompt,
                num_inference_steps=1,
                guidance_scale=0.0).images
    image_list.append(images[0])
for i in range(10):
    image_list[i].save(f"./image{i}.png")
```

image

**Comment by ayushtues** [comment]
Here are some samples

image5image8image9

**Comment by sayakpaul** [comment]
Nice very cool. Feel free to reformat and then we can merge!

**Comment by ayushtues** [comment]
Fixed formatting, and added some documentation!

File: examples/community/README.md [hunk 1]
```diff
@@ -3054,3 +3055,24 @@ output_image = PIL.Image.fromarray(output)
...
+### Instaflow Pipeline
+InstaFlow is an ultra-fast, one-step image generator that achieves image quality close to Stable Diffusion, significantly reducing the demand of computational resources. This efficiency is made possible through a recent [Rectified Flow](https://github.com/gnobitab/RectifiedFlow) technique, which trains probability flows with straight trajectories, hence inherently requiring only a single step for fast inference.
+
+```python
+from diffusers import DiffusionPipeline
+import torch
+
+
+pipe = DiffusionPipeline.from_pretrained("XCLIU/instaflow_0_9B_from_sd_1_5", torch_dtype=torch.float32, custom_pipeline="instaflow_one_step")
+pipe.do_lora() ### use dreambooth lora for better quality
```
Comment (sayakpaul) [inline examples/community/README.md:3068]: Can't we do `load_lora_weights()` here? Let's try use that instead.

**Review by sayakpaul — COMMENTED** [review]
One comment.

**Comment by ayushtues** [comment]
Hi @sayakpaul  the current code (https://github.com/huggingface/diffusers/blob/8793472e83bf309ce18be450d7013b16dfe0ea88/examples/community/instaflow_one_step.py#L433) does the below, using DreamShaper for LoRA (https://huggingface.co/Lykon/dreamshaper-7/tree/main)

- unet = instaflow unet + (dreamshaper unet - SD unet)
- VAE  = dreamshaper VAE
- Encoder = dreamshaper Encoder

Is there a way to do this using `load_lora_weights`?

**Comment by sayakpaul** [comment]
If the underlying LoRA format is new, we cannot assume `load_lora_weights()` will work right off the bat.

**Comment by patrickvonplaten** [comment]
Good to merge for me whenever!

**Comment by ayushtues** [comment]
@sayakpaul since the text encoder and VAE are just copies of the DreamShaper model and the Unet is following a different strategy, I am not sure how to do this using just `load_lora_weights`

**Comment by sayakpaul** [comment]
The UNet doesn't have different pre-trained weights right? 

So, if you subclass the UNet with `UNet2DConditionLoadersMixin` ([example](https://github.com/huggingface/diffusers/blob/17546020fc0e7a09e367540ac578e26e49a2e8be/src/diffusers/models/unet_2d_condition.py#L71)) and the pipeline with `LoraLoaderMixin` ([example](https://github.com/huggingface/diffusers/blob/17546020fc0e7a09e367540ac578e26e49a2e8be/src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py#L119); for SDXL, it should `SDXLLoraLoaderMixin`), that should cut it until and unless I am missing out on something utterly obvious.

**Comment by ayushtues** [comment]
Small update here, I digged a bit into it, and the VAE & Encoder are the same b/w DreamShaper and Instaflow, probably just the standard SD ones, so just need to figure out LoRA for the Unet

**Comment by ayushtues** [comment]
@sayakpaul is there a breaking change to `save_attn_procs` recently, was trying it out for saving the Unet and wasn't working. 

Doesn't seem to be working for a simple example also, 

```python
import torch
from diffusers import DiffusionPipeline

pipeline = DiffusionPipeline.from_pretrained(
    "CompVis/stable-diffusion-v1-4",
    torch_dtype=torch.float32,
)
pipeline.unet.save_attn_procs("./", weight_name="pytorch_custom_diffusion_weights.bin")
```
Gives the error

```
TypeError                                 Traceback (most recent call last)
[<ipython-input-2-532e5d8c7935>](https://localhost:8080/#) in <cell line: 8>()
      6     torch_dtype=torch.float32,
      7 )
----> 8 pipeline.unet.save_attn_procs("./", weight_name="pytorch_custom_diffusion_weights.bin")

5 frames
[/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py](https://localhost:8080/#) in add_module(self, name, module)
    608         """
    609         if not isinstance(module, Module) and module is not None:
--> 610             raise TypeError(f"{torch.typename(module)} is not a Module subclass")
    611         elif not isinstance(name, str):
    612             raise TypeError(f"module name should be a string. Got {torch.typename(name)}")

TypeError: diffusers.models.attention_processor.AttnProcessor2_0 is not a Module subclass
```

Tried both the stable version of diffusers and building from source, colab here - https://colab.research.google.com/drive/1v9i09Mi9L93Cts0zywEj10t1rJVCuJmj?usp=sharing

**Comment by sayakpaul** [comment]
`save_attn_procs()` is deprecated.

**Comment by ayushtues** [comment]
Is there any new functionality for similar usecases, namely saving weights which we can load using `load_attn_procs`?

**Comment by sayakpaul** [comment]
You can use `load_lora_weights()`. The weights should be serialized via `LoraLoaderMixin.save_lora_weights()`. Here's a concrete example:

https://github.com/huggingface/diffusers/blob/dadd55fb36acc862254cf935826d54349b0fcd8c/examples/dreambooth/train_dreambooth_lora_sdxl.py#L1673

**Comment by ayushtues** [comment]
While I am yet to figure out how to use the Dreamshaper model using LORA (it doesn't seem to be using a trivial lora format), the pipeline does seem to be able to use other LORAs using simply `load_lora_weights`. An example with https://huggingface.co/artificialguybr/logo-redmond-1-5v-logo-lora-for-liberteredmond-sd-1-5

```python

import torch
from examples.community.instaflow_one_step import InstaFlowPipeline


pipe = InstaFlowPipeline.from_pretrained("XCLIU/instaflow_0_9B_from_sd_1_5", torch_dtype=torch.float32, cache_dir="E:/InstaFlow/code/cache", requires_safety_checker=False) 
pipe.load_lora_weights("artificialguybr/logo-redmond-1-5v-logo-lora-for-liberteredmond-sd-1-5")
prompt = "logo, A logo for a fitness app, dynamic running figure, energetic colors (red, orange) ),LogoRedAF ,"
generator = torch.Generator(device="cpu").manual_seed(0)
images = pipe(prompt=prompt, 
              num_inference_steps=1,
              generator=generator, 
              guidance_scale=0.0).images 
images[0].save("./image1.png")
```

Without LORA
image1

WITH LORA
image0

**Comment by sayakpaul** [comment]
Nice. Let's include this in the README. And then ship!

**Comment by ayushtues** [comment]
Maybe we can merge the pipeline without the dreamshaper LORA for now, and I can work on adding that support later in another PR

File: examples/community/README.md [hunk 2]
```diff
@@ -3149,6 +3150,43 @@ output_image.save("./output.png")
...
 
+### Instaflow Pipeline
+InstaFlow is an ultra-fast, one-step image generator that achieves image quality close to Stable Diffusion, significantly reducing the demand of computational resources. This efficiency is made possible through a recent [Rectified Flow](https://github.com/gnobitab/RectifiedFlow) technique, which trains probability flows with straight trajectories, hence inherently requiring only a single step for fast inference.
+
+```python
+from diffusers import DiffusionPipeline
+import torch
+
+
+pipe = DiffusionPipeline.from_pretrained("XCLIU/instaflow_0_9B_from_sd_1_5", torch_dtype=torch.float32, custom_pipeline="instaflow_one_step")
```
Comment (sayakpaul) [inline examples/community/README.md:3161]: Mixed-precision doesn't work?

File: examples/community/README.md [hunk 3]
```diff
@@ -3149,6 +3150,43 @@ output_image.save("./output.png")
...
+pipe.to("cuda")  ### if GPU is not available, comment this line
+pipe.load_lora_weights("artificialguybr/logo-redmond-1-5v-logo-lora-for-liberteredmond-sd-1-5")
+prompt = "logo, A logo for a fitness app, dynamic running figure, energetic colors (red, orange) ),LogoRedAF ,"
+images = pipe(prompt=prompt,
+            num_inference_steps=1,
+            guidance_scale=0.0).images
+images[0].save("./image.png")
+```
+![image0](https://github.com/ayushtues/diffusers/assets/43698245/e896a42f-cdae-40bc-9307-947e6c0636c4)
```
Comment (sayakpaul) [inline examples/community/README.md:3188]: nit: You can host these assets on the Hub platform to keep the code repositories clean.

**Comment by ayushtues** [comment]
Made the changes, one small thing, I haven't yet tested directly loading using `DiffusionPipeline`, my current understanding is that it requires the PR to be merged, that is the case right?

**Comment by sayakpaul** [comment]
> Made the changes, one small thing, I haven't yet tested directly loading the custom pipeline using DiffusionPipeline, my current understanding is that it requires the PR to be merged, that is the case right?

Hoe have you tested then? Ccing @patrickvonplaten for better clarification here.

**Comment by ayushtues** [comment]
Hi @sayakpaul, I tested out the pipeline locally using `DiffusionPipeline` as well, colab here -https://colab.research.google.com/drive/1v9i09Mi9L93Cts0zywEj10t1rJVCuJmj?usp=sharing#scrollTo=2rMIAjT7uGVD

To test it locally, we need to create a new folder, with a `pipeline.py` file containing the code of the pipeline, in this case I made a `instaflow_pipe` folder containing a pipeline.py and it worked!

```python
import torch
from diffusers import DiffusionPipeline

pipe = DiffusionPipeline.from_pretrained("XCLIU/instaflow_0_9B_from_sd_1_5", custom_pipeline="./instaflow_pipe")

```

After the PR is merged, `from_pretrained` of `DiffusionPipeline` picks up the code from the community example folder in the main branch of diffusers from github, 
 as can be seen here in the comments - https://github.com/huggingface/diffusers/blob/d3eb3b35bec35f3f4e56b7528577714bf489d464/src/diffusers/pipeline_utils.py#L248C1-L277C31

so then we can do

```python
import torch
from diffusers import DiffusionPipeline

pipe = DiffusionPipeline.from_pretrained("XCLIU/instaflow_0_9B_from_sd_1_5", custom_pipeline="instaflow_one_step")

```

**Comment by ayushtues** [comment]
Let me know if anything else is needed, else we can merge the PR!

**Comment by sayakpaul** [comment]
Once the CI is green will merge. Thanks for contributions!

**Pull request merged**